src/mscorlib/shared/System/StringComparer.cs

```diff
@@ -276,5 +276,11 @@ public override int GetHashCode(string obj)
         // Equals/GetHashCode methods for the comparer itself. 
         public override bool Equals(object obj) => obj is OrdinalIgnoreCaseComparer;
         public override int GetHashCode() => nameof(OrdinalIgnoreCaseComparer).GetHashCode();
+
+        public void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            info.SetType(typeof(OrdinalComparer));
+            info.AddValue("_ignoreCase", true);
```

Comment:
I think that the best way to deal with this is to add back the bool `_ignoreCase` back to OrdinalComparer. It will make the OrdinalComparer serialize perfectly. And then do the performance optimization that saved the if-check by inheriting from the `OrdinalCompared`.